File: app/src/main/java/com/xpf/p2p/activity/GestureVerifyActivity.java
package com.xpf.p2p.activity;

import android.app.Activity;
import android.content.Context;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.text.Html;
import android.text.TextUtils;
import android.view.View;
import android.view.animation.Animation;
import android.view.animation.AnimationUtils;
import android.widget.FrameLayout;
import android.widget.ImageView;
import android.widget.RelativeLayout;
import android.widget.TextView;
import android.widget.Toast;

import com.atguigu.gesturelock.widget.GestureContentView;
import com.atguigu.gesturelock.widget.GestureDrawline;
import com.xpf.p2p.R;

/**
 * Created by xpf on 2016/11/11 :)
 * Wechat:18091383534
 * Function:手势解锁页面
 * {@link # https://github.com/xinpengfei520/P2P}
 */
public class GestureVerifyActivity extends Activity implements View.OnClickListener {
    /**
     * 手机号码
     */
    public static final String PARAM_PHONE_NUMBER = "PARAM_PHONE_NUMBER";
    /**
     * 意图
     */
    public static final String PARAM_INTENT_CODE = "PARAM_INTENT_CODE";
    private RelativeLayout mTopLayout;
    private TextView mTextTitle;
    private TextView mTextCancel;
    private ImageView mImgUserLogo;
    private TextView mTextPhoneNumber;
    private TextView mTextTip;
    private FrameLayout mGestureContainer;
    private GestureContentView mGestureContentView;
    private TextView mTextForget;
    private TextView mTextOther;
    private String mParamPhoneNumber;
    private long mExitTime = 0;
    private int mParamIntentCode;
    private SharedPreferences mSharedPreferences;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        setContentView(R.layout.activity_gesture_verify);

        ObtainExtraData();

        setUpViews();

        setUpListeners();
    }

    private void ObtainExtraData() {
        mParamPhoneNumber = getIntent().getStringExtra(PARAM_PHONE_NUMBER);
        mParamIntentCode = getIntent().getIntExtra(PARAM_INTENT_CODE, 0);
        mSharedPreferences = this.getSharedPreferences("secret_protect", Context.MODE_PRIVATE);
    }

    private void setUpViews() {
        mTopLayout = (RelativeLayout) findViewById(R.id.top_layout);
        mTextTitle = (TextView) findViewById(R.id.text_title);
        mTextCancel = (TextView) findViewById(R.id.text_cancel);
        mImgUserLogo = (ImageView) findViewById(R.id.user_logo);
        mTextPhoneNumber = (TextView) findViewById(R.id.text_phone_number);
        mTextTip = (TextView) findViewById(R.id.text_tip);
        mGestureContainer = (FrameLayout) findViewById(R.id.gesture_container);
        mTextForget = (TextView) findViewById(R.id.text_forget_gesture);
        mTextOther = (TextView) findViewById(R.id.text_other_account);

        String inputCode = mSharedPreferences.getString("inputCode", "1235789");
        // 初始化一个显示各个点的viewGroup
        mGestureContentView = new GestureContentView(this, true, inputCode,
                new GestureDrawline.GestureCallBack() {

                    @Override
                    public void onGestureCodeInput(String inputCode) {

                    }

                    @Override
                    public void checkedSuccess() {

                        mGestureContentView.clearDrawlineState(0L);

                        Toast.makeText(GestureVerifyActivity.this, "密码正确", Toast.LENGTH_SHORT).show();

                        GestureVerifyActivity.this.finish();
                    }

                    @Override
                    public void checkedFail() {

                        mGestureContentView.clearDrawlineState(1300L);
                        mTextTip.setVisibility(View.VISIBLE);
                        mTextTip.setText(Html.fromHtml("<font color='#c70c1e'>密码错误</font>"));

                        // 左右移动动画
                        Animation shakeAnimation = AnimationUtils.loadAnimation(GestureVerifyActivity.this, R.anim.shake);
                        mTextTip.startAnimation(shakeAnimation);
                    }
                });

        // 设置手势解锁显示到哪个布局里面
        mGestureContentView.setParentView(mGestureContainer);
    }

    private void setUpListeners() {
        mTextCancel.setOnClickListener(this);
        mTextForget.setOnClickListener(this);
        mTextOther.setOnClickListener(this);
    }

    private String getProtectedMobile(String phoneNumber) {

        if (TextUtils.isEmpty(phoneNumber) || phoneNumber.length() < 11) {
            return "";
        }

        StringBuilder builder = new StringBuilder();
        builder.append(phoneNumber.subSequence(0, 3));
        builder.append("****");
        builder.append(phoneNumber.subSequence(7, 11));

        return builder.toString();
    }


    @Override
    public void onClick(View v) {

        switch (v.getId()) {
            case R.id.text_cancel:
                this.finish();
                break;

            default:
                break;
        }
    }
}

File: app/src/main/java/com/xpf/p2p/activity/GestureVerifyActivity.kt
package com.xpf.p2p.activity

import android.app.Activity
import android.content.Context
import android.os.Bundle
import android.text.Html
import android.text.TextUtils
import android.view.View
import android.view.animation.AnimationUtils
import android.widget.Toast
import com.xpf.p2p.R
import com.xpf.p2p.databinding.ActivityGestureVerifyBinding
import com.xpf.p2p.widget.GestureContentView
import com.xpf.p2p.widget.GestureDrawline

/**
 * Created by xpf on 2016/11/11 :)
 * Function:手势解锁页面
 */
class GestureVerifyActivity : Activity() {

    private lateinit var binding: ActivityGestureVerifyBinding
    private lateinit var mGestureContentView: GestureContentView
    private lateinit var mSharedPreferences: android.content.SharedPreferences

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityGestureVerifyBinding.inflate(layoutInflater)
        setContentView(binding.root)
        obtainExtraData()
        setUpViews()
    }

    private fun obtainExtraData() {
        mSharedPreferences = getSharedPreferences("secret_protect", Context.MODE_PRIVATE)
    }

    private fun setUpViews() {
        binding.textCancel.setOnClickListener { finish() }

        val inputCode = mSharedPreferences.getString("inputCode", "1235789") ?: "1235789"
        mGestureContentView = GestureContentView(this, true, inputCode,
            object : GestureDrawline.GestureCallBack {
                override fun onGestureCodeInput(inputCode: String) {}

                override fun checkedSuccess() {
                    mGestureContentView.clearDrawlineState(0L)
                    Toast.makeText(this@GestureVerifyActivity, "密码正确", Toast.LENGTH_SHORT).show()
                    finish()
                }

                override fun checkedFail() {
                    mGestureContentView.clearDrawlineState(1300L)
                    binding.textTip.visibility = View.VISIBLE
                    @Suppress("DEPRECATION")
                    binding.textTip.text = Html.fromHtml("<font color='#c70c1e'>密码错误</font>")
                    val shakeAnimation = AnimationUtils.loadAnimation(this@GestureVerifyActivity, R.anim.shake)
                    binding.textTip.startAnimation(shakeAnimation)
                }
            })

        mGestureContentView.setParentView(binding.gestureContainer)
    }

    private fun getProtectedMobile(phoneNumber: String?): String {
        if (TextUtils.isEmpty(phoneNumber) || phoneNumber!!.length < 11) {
            return ""
        }
        return "${phoneNumber.subSequence(0, 3)}****${phoneNumber.subSequence(7, 11)}"
    }

    companion object {
        const val PARAM_PHONE_NUMBER = "PARAM_PHONE_NUMBER"
        const val PARAM_INTENT_CODE = "PARAM_INTENT_CODE"
    }
}
